tests/unit-core/jerry-ext/jerry-context/jerry-context-test.c

```diff
@@ -0,0 +1,68 @@
+/* Copyright JS Foundation and other contributors, http://js.foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "jerryscript.h"
+#include "test-common.h"
+#include "jerry-context.h"
+
+static char *static_slot1 = "static slot 1";
+static char *static_slot2 = "static slot 2";
+
+static void *
+init_slot1 (void)
+{
+  return static_slot1;
+} /* init_slot1 */
+
+static void
+deinit_slot1 (void *slot)
+{
+  TEST_ASSERT (slot == static_slot1);
+} /* deinit_slot1 */
+
+static void *
+init_slot2 (void)
+{
+  return static_slot2;
+} /* init_slot2 */
+
+static void
+deinit_slot2 (void *slot)
+{
+  TEST_ASSERT (slot == static_slot2);
+} /* deinit_slot2 */
+
+int
+main (int argc, char **argv)
+{
+  (void) argc;
+  (void) argv;
+  TEST_INIT ();
+
+  int slot1 = jerryx_context_request_slot (init_slot1, deinit_slot1);
+  int slot2 = jerryx_context_request_slot (init_slot2, deinit_slot2);
+  int slot3 = jerryx_context_request_slot (init_slot2, deinit_slot2);
+
+  TEST_ASSERT (slot1 == 0);
+  TEST_ASSERT (slot2 == 1);
+  TEST_ASSERT (slot3 == -1);
+
+  jerry_init_with_user_context (JERRY_INIT_EMPTY, jerryx_context_init, jerryx_context_deinit);
```

Comment:
`jerry_init_with_user_context ()/jerry_get_user_context ()` works well in #1778 

I changed the unit-test-user-context, by building it with `./tools/build.py --unittest --jerry-libc=off --external-context=on`, then run `./build/bin/unit-test-user-context`, it passed

I still consider where to put the test case.
``` C
#include "config.h"
#include "jerryscript.h"
#include "test-common.h"
#include "jerryscript-port.h"
#include "jerryscript-port-default.h"


#define JERRY_EXTERNAL_HEAP_SIZE (1024)
#define JERRY_EXTERNAL_LCACHE_ROW_COUNT 4
#define JERRY_EXTERNAL_LCACHE_ROW_LENGTH 2
#define JERRY_EXTERNAL_HEAP_ALIGN 8

static jerry_instance_t *instance1_p;
static jerry_instance_t *instance2_p;

static const char *identifying_string1 = "identifying string 1";
static bool user_context1_new_was_called = false;
static bool user_context1_free_was_called = false;

static const char *identifying_string2 = "identifying string 2";
static bool user_context2_new_was_called = false;
static bool user_context2_free_was_called = false;

static void *
user_context1_new (void)
{
  user_context1_new_was_called = true;
  return (void *) identifying_string1;
} /* user_context1_new */

static void
user_context1_free (void *user_context_p)
{
  user_context1_free_was_called = true;
  TEST_ASSERT (((const char *) user_context_p) == identifying_string1);
} /* user_context1_free */

static void *
user_context2_new (void)
{
  user_context2_new_was_called = true;
  return (void *) identifying_string2;
} /* user_context2_new */

static void
user_context2_free (void *user_context_p)
{
  user_context2_free_was_called = true;
  TEST_ASSERT (((const char *) user_context_p) == identifying_string2);
} /* user_context2_free */

int
main (void)
{
  TEST_INIT ();

  size_t context_size = jerry_get_context_size ();
  size_t lcache_size = jerry_get_lcache_size (JERRY_EXTERNAL_LCACHE_ROW_COUNT, JERRY_EXTERNAL_LCACHE_ROW_LENGTH);
  uintptr_t alignment_mask = ~(uintptr_t) (JERRY_EXTERNAL_HEAP_ALIGN - 1);
  /* create instance 1 */
  void *context1_p = malloc (context_size);
  memset (context1_p, 0, context_size);
  void *lcache1_p = malloc (lcache_size);
  void *heap1_p = malloc (JERRY_EXTERNAL_HEAP_SIZE + JERRY_EXTERNAL_HEAP_ALIGN - 1);
  instance1_p = (jerry_instance_t *) malloc (sizeof (jerry_instance_t));
  instance1_p->context_p = context1_p;
  instance1_p->lcache_p = lcache1_p;
  instance1_p->heap_p = (void *) (((uintptr_t) heap1_p + JERRY_EXTERNAL_HEAP_ALIGN - 1) & alignment_mask);
  instance1_p->heap_size = JERRY_EXTERNAL_HEAP_SIZE;
  instance1_p->lcache_row_count = JERRY_EXTERNAL_LCACHE_ROW_COUNT;
  instance1_p->lcache_row_length = JERRY_EXTERNAL_LCACHE_ROW_LENGTH;
  /* create instance 2 */
  void *context2_p = malloc (context_size);
  memset (context2_p, 0, context_size);
  void *lcache2_p = malloc (lcache_size);
  void *heap2_p = malloc (JERRY_EXTERNAL_HEAP_SIZE + JERRY_EXTERNAL_HEAP_ALIGN - 1);
  instance2_p = (jerry_instance_t *) malloc (sizeof (jerry_instance_t));
  instance2_p->context_p = context2_p;
  instance2_p->lcache_p = lcache2_p;
  instance2_p->heap_p = (void *) (((uintptr_t) heap2_p + JERRY_EXTERNAL_HEAP_ALIGN - 1) & alignment_mask);
  instance2_p->heap_size = JERRY_EXTERNAL_HEAP_SIZE;
  instance2_p->lcache_row_count = JERRY_EXTERNAL_LCACHE_ROW_COUNT;
  instance2_p->lcache_row_length = JERRY_EXTERNAL_LCACHE_ROW_LENGTH;
  /* switch to instance 1 */
  jerry_port_set_instance (instance1_p);
  jerry_init_with_user_context (JERRY_INIT_EMPTY, user_context1_new, user_context1_free);
  TEST_ASSERT ((((const char *)(jerry_get_user_context ()))) == identifying_string1);
  /* switch to instance 2 */
  jerry_port_set_instance (instance2_p);
  jerry_init_with_user_context (JERRY_INIT_EMPTY, user_context2_new, user_context2_free);
  TEST_ASSERT ((((const char *)(jerry_get_user_context ()))) == identifying_string2);
  /* switch to instance 1 */
  jerry_port_set_instance (instance1_p);
  jerry_cleanup ();
  TEST_ASSERT (user_context1_new_was_called);
  TEST_ASSERT (user_context1_free_was_called);
  /* switch to instance 2 */
  jerry_port_set_instance (instance2_p);
  jerry_cleanup ();
  TEST_ASSERT (user_context2_new_was_called);
  TEST_ASSERT (user_context2_free_was_called);

  return 0;
} /* main */

```